### PR TITLE
DAOS-4898 obj: add csum error handling for EC obj fetch

### PR DESCRIPTION
**Pull request by liuxuezhao**
Add csum error handling for EC obj fetch
   When gets csum error for EC obj fetch, it can try to re-fetch it by
   degraded fetch protocol, rather than return error to user.
   Adds related test cases to daos_checksum.c.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Review by wangdi1 — COMMENTED** [review]
the whole retry/degraded/csum state machine is a bit hard to follow, 

Probably obj_ec_get_degrade() return value needs to be check, otherwise will it get into endless lop if it return -DER_DATA_LOSS? Or I miss sth?

File: src/object/cli_obj.c
Comment (liuxuezhao) [reply src/object/cli_obj.c:3427]: OK, sure.

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c:253]: OK, added.

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c:364]: now we only set "shard_arg->reasb_req" for EC obj, but later probably also set for replica obj.

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c:371]: ok, added

File: src/object/cli_obj.c
Comment (liuxuezhao) [reply src/object/cli_obj.c:760]: let's discuss it offline can refine it in following patch, thx.

File: src/object/cli_shard.c
Comment (wangdi1) [reply src/object/cli_shard.c]: I thought reasb_req is only for EC object? you mean some ec object does not need reasb or  reasb is also needed for replication object?

File: src/object/cli_shard.c
Comment (wangdi1) [inline src/object/cli_shard.c:258]: If rc == -DER_DATA_LOSS, should the return value is still DER_CSUM? which will still cause retry?

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c]: now reasb_req is only for EC obj. But later possibly also for replica obj, for example if need to do the re-order of recx etc.

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c:258]: when there is no more available tgt, the retry will fail immediately.
for that case, user expected to get ECSUM error code.

Comment (wangdi1) [reply src/object/cli_shard.c:258]: ah, I see. thanks.

**Review by wangdi1 — APPROVED** [review]

**Comment by liuxuezhao** [comment]
ping ...

**Review by wangdi1 — DISMISSED** [review]

**Comment by liuxuezhao** [comment]
just a rebase to resolve confliction

**Review by wangdi1 — APPROVED** [review]

**Comment by liuxuezhao** [comment]
rebase due to conflict again

**Review by wangdi1 — DISMISSED** [review]

File: src/object/cli_shard.c
Comment (ryon-jensen) [inline src/object/cli_shard.c:235]: Don't think the curly braces are needed

File: src/common/checksum.c
Comment (ryon-jensen) [inline src/common/checksum.c:690]: Remind me what it means when singv_idx == -1?

File: src/common/checksum.c
Comment (ryon-jensen) [inline src/common/checksum.c:692]: csum_buf_len is confusing to me. Sounds like it's the buffer length of the checksums, but looks like it is the number of bytes of the data that will be used to calculate the checksum? Is that right?

**Review by ryon-jensen — COMMENTED** [review]
I'm not as familiar with the EC logic and the calc_csum_sv function is somewhat difficult for me to follow. Can you add more comments or abstract out some of the logic to make it easier to follow when the different conditional paths will be hit?

File: src/common/checksum.c
Comment (liuxuezhao) [reply src/common/checksum.c:690]: -1 means calculate all shards' checksum for evenly-distributed single-value.

there are some comments for daos_csummer_calc_iods() and daos_csummer_verify_iod() in src/include/daos/checksum.h about it:
295  * @param singv_idx>    single value target index, valid when singv_los         
296  *>     >       >       is non-NULL. -1 means verifying csum for all shards.

File: src/common/checksum.c
Comment (liuxuezhao) [reply src/common/checksum.c:692]: correct. hmm, agreed the name looks confuse, let me rename it...

File: src/object/cli_shard.c
Comment (liuxuezhao) [reply src/object/cli_shard.c:235]: right.

**Comment by liuxuezhao** [comment]
rebase + address ryon's comments + add another small patch to simply csum_report handling (to save CI resource).

**Review by wangdi1 — APPROVED** [review]

File: src/common/checksum.c
Comment (ryon-jensen) [reply src/common/checksum.c:690]: Ah, thanks.

**Review by ryon-jensen — APPROVED** [review]

**Comment by liuxuezhao** [comment]
"All tests are passing
Nice one! All 3632 tests for this pipeline are passing."

**Comment by jolivier23** [comment]
> "All tests are passing
> Nice one! All 3632 tests for this pipeline are passing."

Unfortunately, this is due to being based on a version that had disabled several tests

**Comment by liuxuezhao** [comment]
> > "All tests are passing
> > Nice one! All 3632 tests for this pipeline are passing."
> 
> Unfortunately, this is due to being based on a version that had disabled several tests

just rebased to retrigger

**Review by wangdi1 — DISMISSED** [review]

**Review by ryon-jensen — DISMISSED** [review]

**Comment by liuxuezhao** [comment]
the CI failures due to https://jira.hpdd.intel.com/browse/DAOS-5919 (DER_OOG in pool create).

**Comment by liuxuezhao** [comment]
@wangdi1  @ryon-jensen @jgmoore-or ping, can you approve it again? just rebasing after your last approval.

**Review by wangdi1 — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]

**Pull request merged**
